### PR TITLE
Grounder modularization

### PR DESCRIPTION
**Pull request by lorenzleutgeb**
I resolved (3.) from #72 and refactored `NaiveGrounder` a bit.

It now has a `NogoodRegistry`, a `WorkingMemory` and a `ChoiceRecorder`.

Also, I decided to split `FixedInterpretation` from `Predicate`. All the interpretation classes that correspond to lambdas or methods are now more independent from `Predicate`.

~~I also introduced `SymbolWithRank` which is used for `Predicate` (predicate symbol and its arity), as well as `FunctionSymbol`, which in turn divides into `Functor` for "real" function names and `Constant` for nullary function symbols.~~ (removed after discussion)

### CODE REVIEW

**Comment by AntoniusW** [comment]
Now this is bad timing, since I am also in the middle of several changes to the NaiveGrounder in order to deal with variables bindings correctly. I will open another PR soon, then we can decide which one is easier to put on top of the other. You addressed several other issues than I do right now, so maybe it is not that complicated. :-)

**Comment by lorenzleutgeb** [comment]
OK, then I'll just leave this aside for a moment and we consolidate afterwards.

**Comment by AntoniusW** [comment]
With the changes on variable binding merged into master, this should no longer be blocked. Especially the newly introduced `WorkingMemory` and `ChoiceRecorder` look useful.
In light of #87 the `NoGoodRegistry` is not necessary and the IDs of NoGoods could be removed from Alpha altogether (maybe in another PR or directly here).

**Comment by lorenzleutgeb** [comment]
Yes, I will soon update this PR.

**Comment by lorenzleutgeb** [comment]
I updated this branch and think that these changes really improve the "code quality" of the grounder.

@AntoniusW please reconsider.

File: src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java [hunk 1]
```diff
@@ -94,8 +89,8 @@ public NaiveGrounder(Program program, java.util.function.Predicate<Predicate> fi
 		for (Atom fact : program.getFacts()) {
 			final Predicate predicate = fact.getPredicate();
 
-			// Record predicate
-			adaptWorkingMemoryForPredicate(predicate);
+			// Record interpretation
```
Comment (AntoniusW) [inline src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java]: "predicate." I think.

Comment (lorenzleutgeb) [reply src/main/java/at/ac/tuwien/kr/alpha/grounder/NaiveGrounder.java]: Yes. Sorry.

**Comment by AntoniusW** [comment]
One of the tests still fails, would merge otherwise.

**Comment by lorenzleutgeb** [comment]
Oh, this seems to be a side effect after merging #96 ... will fix.

**Comment by lorenzleutgeb** [comment]
I took the liberty to merge this after your positive comment.

**Comment by AntoniusW** [comment]
This was my intention when writing that comment.

**Pull request merged**